### PR TITLE
Optimize configuration management services

### LINKED ISSUE #1622: Optimize configuration management services
### What would you like to be enhanced?

1、Modify the configuration management function and render the configuration page based on the template

### Why is this needed?

1、Decoupling configuration management functionality from plugin configuration

### PR DESCRIPTION
**Pull request by hanbingleixue**
**What type of PR is this?**

Enhancement

**What this PR does / why we need it?**

1. Modify configuration query interface
2. Modify the configuration details page and render according to the configuration template
image
image

3. Add a configuration template query interface

**Which issue(s) this PR fixes？**

Fixes #1622

**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [ ] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java [hunk 1]
```diff
@@ -231,14 +232,15 @@ public Map<String, List<String>> getConfigList(String key, String group, boolean
...
+        Pattern pattern = Pattern.compile(covertGroup);
         for (KieConfigEntity entity : kieResponse.getData()) {
             if (exactMatchFlag) {
                 List<String> configList = result.computeIfAbsent(LabelGroupUtils.createLabelGroup(entity.getLabels()),
                         configKey -> new ArrayList<>());
                 configList.add(entity.getKey());
             } else {
                 String currentConfigGroup = LabelGroupUtils.createLabelGroup(entity.getLabels());
-                if (currentConfigGroup.contains(covertGroup)) {
+                if (currentConfigGroup.contains(covertGroup) || pattern.matcher(covertGroup).matches()) {
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java]: pattern is getted by covertGroup

File: sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Template controller
+ *
+ * @author zhp
+ * @since 2024-08-22
+ */
+@RestController
+@RequestMapping("/sermant")
+public class TemplateController {
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java]: what kind of template

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 3]
```diff
@@ -69,66 +68,56 @@ public class ConfigService {
...
-    private static final ScheduledExecutorService EXECUTOR_SERVICE = Executors.newScheduledThreadPool(1);
-
     /**
      * ZK authorization separator
      */
     private static final char ZK_AUTH_SEPARATOR = ':';
 
     private static final String SCHEME = "digest";
 
+    private final Pattern pattern = Pattern.compile("\\$\\{\\s*\\w+\\s*\\}");
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: describe this parttern‘s usage

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 4]
```diff
@@ -69,66 +68,56 @@ public class ConfigService {
...
-                group = group.replace(KieConstants.DEFAULT_LABEL_PRE, StringUtils.EMPTY);
-            }
-            if (!exactMatchFlag && !handler.verifyConfigurationGroup(group)) {
-                continue;
-            }
-            for (String configKey : entry.getValue()) {
-                if (!exactMatchFlag && !handler.verifyConfigurationKey(configKey)) {
-                    continue;
+        groupRules.forEach(groupRule -> {
+            String groupPattern = replaceWithWildcard(groupRule);
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: replacexxxWithWildcard

File: sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+import javax.annotation.Resource;
+
+/**
+ * Template controller
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java]: describe controller

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java [hunk 6]
```diff
@@ -231,14 +232,15 @@ public Map<String, List<String>> getConfigList(String key, String group, boolean
             return Collections.emptyMap();
         }
         Map<String, List<String>> result = new HashMap<>();
+        Pattern pattern = Pattern.compile(covertGroup);
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java]: convertedGroup

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java [hunk 7]
```diff
@@ -193,7 +198,8 @@ private void fillChildrenInfo(String path, Map<String, List<String>> configMap,
                 configMap.put(childPath.substring(1), subChild);
                 return;
             }
-            List<String> matchSubChild = subChild.stream().filter(value -> value.contains(key))
+            List<String> matchSubChild = subChild.stream()
+                    .filter(value -> value.contains(key) || Pattern.compile(key).matcher(value).matches())
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: Pattern.compile(key) can be extrated

File: sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java [hunk 8]
```diff
@@ -139,14 +139,14 @@ public class CommonConst {
                     + "---  " + System.lineSeparator();
 
     /**
-     * Service name for global configuration
+     * Plugin type for non-Semant plugins
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java]: Why do non-sermant plugins exist?

File: sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java [hunk 9]
```diff
@@ -139,14 +139,14 @@ public class CommonConst {
...
      */
-    public static final String GLOBAL_CONFIGURATION_SERVICE_NAME = "ALL_SERVICE";
+    public static final String OTHER_PLUGIN = "other";
 
     /**
-     * Default regular expression for configuration item Group
+     * Wildcard
      */
-    public static final String CONFIGURATION_DEFAULT_PATTERN = "^(app=[^&]*+&environment=[^&]*+(&service=[^&]*)?)?$";
+    public static final String WILDCARD = "[^&]+";
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java]: WILDCARD should  be *

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 10]
```diff
@@ -69,66 +68,56 @@ public class ConfigService {
...
+            Map<String, List<String>> configMap = client.getConfigList(CommonConst.WILDCARD, groupPattern, false);
+            configMap.forEach((group, value) -> value.forEach(key -> keyRules.forEach(keyRule -> {
+                String keyPattern = replaceWithWildcard(keyRule);
+                if (Pattern.compile(keyPattern).matcher(key).matches()) {
+                    ConfigInfo configInfo = new ConfigInfo(key, group, keyRule, groupRule, request.getPluginType(),
+                            request.getNamespace());
+                    configInfoList.add(configInfo);
                 }
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: extract to a method

File: sermant-backend/src/main/java/io/sermant/backend/service/TemplateService.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * Get the template information
+     *
+     * @param pluginName plugin name
+     * @return the template information
+     */
+    public Result<PluginTemplateInfo> getTemplate(String pluginName) {
+        if (CollectionUtils.isEmpty(pluginTemplateInfoList)) {
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/TemplateService.java]: Improper return value

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 10 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java (hunk 1 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java (hunk 2 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 3 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 4 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java (hunk 5 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/controller/TemplateController.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java (hunk 6 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java (hunk 7 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java (hunk 8 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java (hunk 9 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/common/conf/CommonConst.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/TemplateService.java (hunk 11 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/TemplateService.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java [hunk 12]
```diff
@@ -221,24 +222,25 @@ public String getConfig(String key, String group) {
...
-            kieResponse = getKieResponse(key, covertGroup, exactMatchFlag);
+            kieResponse = getKieResponse(key, convertedGroup, exactMatchFlag);
         } else {
             kieResponse = getKieResponse(key, null, exactMatchFlag);
         }
         if (kieResponse == null || kieResponse.getData() == null) {
             return Collections.emptyMap();
         }
         Map<String, List<String>> result = new HashMap<>();
+        Pattern pattern = Pattern.compile(convertedGroup);
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java]: use a map to cache these Pattern for better query efficiency

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java [hunk 13]
```diff
@@ -175,9 +176,13 @@ private Map<String, List<String>> fuzzyGetConfigListByGroupAndKey(String key, St
     private void fillChildrenInfo(String path, Map<String, List<String>> configMap, String nodeName,
                                   String key) throws InterruptedException, KeeperException {
         List<String> children = this.zkClient.getChildren(path, false);
+        if (children == null || children.isEmpty()) {
+            return;
+        }
+        Pattern groupPattern = Pattern.compile(nodeName);
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: cache patterns like the comment in kieclient

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 14]
```diff
@@ -69,66 +68,66 @@ public class ConfigService {
...
-                ConfigInfo configInfo = handler.parsePluginInfo(configKey, group);
-                if (!exactMatchFlag && !handler.filterConfiguration(request, configInfo)) {
-                    continue;
+        groupRules.forEach(groupRule -> {
+            String groupPattern = replaceVariableWithWildcard(groupRule);
+            Map<String, List<String>> configMap = client.getConfigList(CommonConst.WILDCARD, groupPattern,
+                    request.isExactMatchFlag());
+            configMap.forEach((group, value) -> value.forEach(key -> keyRules.forEach(keyRule -> {
+                String keyPattern = replaceVariableWithWildcard(keyRule);
+                if (Pattern.compile(keyPattern).matcher(key).matches()) {
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: cache patterns

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java (hunk 12 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/kie/client/kie/KieClient.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java (hunk 13 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 14 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

**Review by lilai23 — DISMISSED** [review]

**Review by daizhenyu — DISMISSED** [review]

**Review by lilai23 — APPROVED** [review]

**Review by daizhenyu — APPROVED** [review]

**Pull request merged**